### PR TITLE
Implement reabstraction in CoreCLR for Static Virtual Methods

### LINKED ISSUE #71414: Static virtual reabstraction doesn't seem to work
I'm working on implementing this with NativeAOT and was looking at how it works in the VM.

Reproed with ~~7.0.100-preview.5.22307.18 SDK~~7.0.100-preview.7.22328.2:

Compile following program:

```csharp
Call<BarClass>();

static void Call<T>() where T : IFoo => T.Frob();

interface IFoo
{
    static virtual void Frob() => throw null;
}

interface IBar : IFoo
{
    static void IFoo.Frob() => throw null;
}

interface IBaz : IFoo
{
    static abstract void IFoo.Frob();
}

class BarClass : IBar
{
}
```

~~This will throw at runtime:~~

```
Unhandled exception. System.TypeLoadException: Method implementation on an interface 'IBar' from assembly 'publishaot2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' must be a final method.
   at Program.<Main>$(String[] args)
```

~~Sure, looks like a Roslyn bug, but it's not what I'm after. (@AlekseyTs is this known?)~~

~~Fix the problem by ildasm-ing the assembly, adding the missing `final` and ILAsm-ing it back. Now the program works as expected.~~

Now, go back to the IL and change BarClass to implement IBaz instead of the IBar (Roslyn wouldn't allow compiling that because the method was reabstracted). Rebuild the IL.

```
Unhandled exception. System.MissingMethodException: Method not found: 'Void IFoo.Frob()'.
   at Program.<Main>$(String[] args)
```

This doesn't look like the experience we would want for reabstraction. For reference, the exception with instance default interface methods is (you have to go back to the IL this generates and replace IBar with IBaz in BarClass interface list, same as for the static virtual case because Roslyn doesn't allow compiling with the reabstraction):

```csharp
IFoo foo = new BarClass();
foo.Frob();

interface IFoo
{
    virtual void Frob() => throw null;
}

interface IBar : IFoo
{
    void IFoo.Frob() => throw null;
}

interface IBaz : IFoo
{
    abstract void IFoo.Frob();
}

class BarClass : IBar
{
}
```

This will print a more descriptive message:

```
Unhandled exception. System.EntryPointNotFoundException: Could not call method 'IFoo.Frob()' on type 'IFoo' with an instance of 'BarClass' from assembly 'publishaot2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' because there is no implementation for the method.
   at IFoo.Frob()
   at Program.<Main>$(String[] args)
```

Cc @trylek @davidwrighton

### PR DESCRIPTION
**Pull request by davidwrighton**
- Add a new helper for the re-abstraction case when the JIT detects an issue
- In the late bound case, instead of producing an error directly, produce an IL stub and have it make the virtual stub dispatch call itself, this will fall back to the JIT implementation for re-abstraction
- Tweak the late bound case to also actively detect the AmbiguousMatchException case as well and use the same helper there as well.

Fixes #71414

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
The issue to block the test on for NativeAOT is https://github.com/dotnet/runtime/issues/72589. It didn't make sense to implement if CoreCLR proper doesn't implement it.

File: src/coreclr/vm/ilstubcache.cpp [hunk 1]
```diff
@@ -159,6 +159,7 @@ namespace
             case DynamicMethodDesc::StubWrapperDelegate:    return "IL_STUB_WrapperDelegate_Invoke";
             case DynamicMethodDesc::StubTailCallStoreArgs:  return "IL_STUB_StoreTailCallArgs";
             case DynamicMethodDesc::StubTailCallCallTarget: return "IL_STUB_CallTailCallTarget";
+            case DynamicMethodDesc::StubVirtualStaticMethodDispatch: return "IL_STUB_bVirtualStaticMethodDispatch";
```
Comment (trylek) [inline src/coreclr/vm/ilstubcache.cpp:162]: Nit - is the lowercase 'b' after <code>IL_STUB_</code> intentional or just a typo?

**Review by trylek — APPROVED** [review]
Looks great to me, thanks David!

**Comment by davidwrighton** [comment]
Only failing issue not flagged by BuildAnalysis as known is #88870 which simply isn't being detected correctly by the BuildAnalysis stuff.

**Pull request merged**